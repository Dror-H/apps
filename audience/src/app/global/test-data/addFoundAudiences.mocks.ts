import { SearchResult, SupportedProviders } from '@instigo-app/data-transfer-object';

export const addFoundAudiencesSearchResults1: SearchResult[] = [
  {
    id: '14b4016c-41dc-47cb-85e3-c61ee3650fdc',
    spec: {
      provider: SupportedProviders.FACEBOOK,
      include: {
        and: [
          { or: { ageRange: [{ providerId: '(30,60)', type: 'ageRange' }] } },
          { or: { locales: [{ providerId: [] }] } },
          { or: {} },
        ],
      },
      exclude: { or: {} },
    },
    topics: [],
    rank: 100,
    specRatio: { interests: 100, demographics: 0, behaviors: 0 },
    userTags: [],
  },
  {
    id: '24b4016c-41dc-47cb-85e3-c61ee3650fdg',
    spec: {
      provider: SupportedProviders.FACEBOOK,
      include: {
        and: [
          { or: { ageRange: [{ providerId: '(30,60)', type: 'ageRange' }] } },
          { or: { locales: [{ providerId: [] }] } },
          { or: {} },
        ],
      },
      exclude: { or: {} },
    },
    topics: [],
    rank: 50,
    specRatio: { interests: 100, demographics: 0, behaviors: 0 },
    userTags: [],
  },
];
export const addFoundAudiencesSearchResults2: SearchResult[] = [
  {
    id: '34b4016c-41dc-47cb-85e3-c61ee3651234',
    spec: {
      provider: SupportedProviders.FACEBOOK,
      include: {
        and: [
          { or: { ageRange: [{ providerId: '(30,60)', type: 'ageRange' }] } },
          { or: { locales: [{ providerId: [] }] } },
          { or: {} },
        ],
      },
      exclude: { or: {} },
    },
    topics: [],
    rank: 75,
    specRatio: { interests: 100, demographics: 0, behaviors: 0 },
    userTags: [],
  },
];

export const addFoundAudiencesSortedResult: SearchResult[] = [
  addFoundAudiencesSearchResults1[0],
  addFoundAudiencesSearchResults2[0],
  addFoundAudiencesSearchResults1[1],
];
