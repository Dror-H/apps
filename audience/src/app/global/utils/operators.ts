import { DisplayNotificationService } from '@audience-app/global/services/display-notification/display-notification.service';
import { NotificationType, UINotification } from '@instigo-app/data-transfer-object';
import { MonoTypeOperatorFunction, throwError } from 'rxjs';
import { catchError, tap } from 'rxjs/operators';

export interface ErrorType {
  error?: { message?: string; title?: string; description?: string };
  message?: string;
}

export function notificationOnErrorOperator<T>(
  displayNotificationService: DisplayNotificationService,
  notificationOptions?: Partial<UINotification>,
): MonoTypeOperatorFunction<T> {
  return catchError((err: ErrorType) => {
    const notification = new UINotification({
      type: NotificationType.ERROR,
      content: notificationOptions?.content || err.error?.message || err.error?.description || err.message,
      title: notificationOptions?.title || err?.error?.title,
      options: notificationOptions?.options,
    });
    displayNotificationService.displayNotification(notification);
    return throwError(err);
  });
}
export function notificationOnSuccessOperator<T>(
  displayNotificationService: DisplayNotificationService,
  notificationOptions: Omit<UINotification, 'type'>,
): MonoTypeOperatorFunction<T> {
  return tap(() => {
    const notification = new UINotification({
      ...notificationOptions,
      type: NotificationType.SUCCESS,
    });
    displayNotificationService.displayNotification(notification);
  });
}

export function notificationOperator<T>(
  displayNotificationService: DisplayNotificationService,
  notificationOptions: UINotification,
): MonoTypeOperatorFunction<T> {
  return tap(() => {
    const notification = new UINotification({ ...notificationOptions });
    displayNotificationService.displayNotification(notification);
  });
}
