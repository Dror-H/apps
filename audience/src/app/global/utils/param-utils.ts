export function getArrayFromParams(params?: string | string[]): string[] {
  if (!params || (params && !params.length)) {
    return [];
  }

  return typeof params === 'string' ? [params] : params;
}
