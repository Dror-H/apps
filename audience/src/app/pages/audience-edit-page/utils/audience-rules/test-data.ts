import { AudienceDto, SupportedProviders, TargetingDto } from '@instigo-app/data-transfer-object';

export const mockRegularTargeting: TargetingDto = {
  provider: SupportedProviders.FACEBOOK,
  include: {
    and: [
      { or: { ageRange: [{ providerId: '(25,50)', type: 'ageRange' }] } },
      { or: { locales: [{ providerId: [] }] } },
      {
        or: {
          behaviors: [
            {
              providerId: '6002714898572',
              type: 'behaviors',
              path: ['Behaviours', 'Digital activities', 'Small business owners'],
              name: 'Small business owners',
              audienceSize: 48970322,
            },
          ],
          interests: [
            {
              providerId: '6002884511422',
              type: 'interests',
              path: ['Interests', 'Business and industry', 'Small business'],
              name: 'Small business',
              audienceSize: 192686100.5,
            },
          ],
          employers: [
            {
              providerId: '109288035760640',
              type: 'employers',
              path: ['Demographics', 'Work', 'Employers', 'Business Owner'],
              name: 'Business Owner',
              audienceSize: 116699,
            },
          ],
        },
      },
      {
        or: {
          interests: [
            {
              providerId: '6002964301046',
              type: 'interests',
              path: ['Interests', 'Additional interests', 'Business school'],
              name: 'Business school',
              audienceSize: 41083805,
            },
          ],
          fieldsOfStudy: [
            {
              providerId: '106335129403847',
              type: 'fieldsOfStudy',
              path: ['Demographics', 'Education', 'Fields of study', 'Business management'],
              name: 'Business management',
              audienceSize: 2636510.5,
            },
          ],
          educationStatuses: [{ type: 'educationStatuses', providerId: '' }],
        },
      },
      {
        or: {
          interests: [
            {
              providerId: '6002964301046',
              type: 'interests',
              path: ['Interests', 'Additional interests', 'Business school'],
              name: 'Business school',
              audienceSize: 41083805,
            },
          ],
          employers: [
            {
              providerId: '116050848438791',
              type: 'employers',
              path: ['Demographics', 'Work', 'Employers', 'Currently looking for a job.'],
              name: 'Currently looking for a job.',
              audienceSize: 3339.5,
            },
          ],
        },
      },
    ],
  },
  exclude: {
    or: {
      behaviors: [
        {
          providerId: '6071631541183',
          type: 'behaviors',
          path: ['Behaviours', 'Purchase behaviour', 'Engaged shoppers'],
          name: 'Engaged shoppers',
          audienceSize: 539317289.5,
        },
        {
          providerId: '123456',
          type: 'behaviors',
          path: ['Behaviours'],
          name: 'test',
          audienceSize: 123456,
        },
      ],
      demographics: [
        {
          providerId: '654321',
          type: 'demographics',
          path: ['Demographics'],
          name: 'test 2',
          audienceSize: 654321,
        },
      ],
      interests: [
        {
          providerId: '123654',
          type: 'interests',
          path: ['Interests'],
          name: 'test 2',
          audienceSize: 321654,
        },
      ],
    },
  },
};

export const mockRegularTargetingWithLocation: TargetingDto = {
  provider: SupportedProviders.FACEBOOK,
  include: {
    and: [
      { or: { ageRange: [{ providerId: '(25,50)', type: 'ageRange' }] } },
      { or: { locales: [{ providerId: [] }] } },
      { or: { locationTypes: [], countries: [] } },
      {
        or: {
          behaviors: [
            {
              providerId: '6002714898572',
              type: 'behaviors',
              path: ['Behaviours', 'Digital activities', 'Small business owners'],
              name: 'Small business owners',
              audienceSize: 48970322,
            },
          ],
          interests: [
            {
              providerId: '6002884511422',
              type: 'interests',
              path: ['Interests', 'Business and industry', 'Small business'],
              name: 'Small business',
              audienceSize: 192686100.5,
            },
          ],
          employers: [
            {
              providerId: '109288035760640',
              type: 'employers',
              path: ['Demographics', 'Work', 'Employers', 'Business Owner'],
              name: 'Business Owner',
              audienceSize: 116699,
            },
          ],
        },
      },
      {
        or: {
          interests: [
            {
              providerId: '6002964301046',
              type: 'interests',
              path: ['Interests', 'Additional interests', 'Business school'],
              name: 'Business school',
              audienceSize: 41083805,
            },
          ],
          fieldsOfStudy: [
            {
              providerId: '106335129403847',
              type: 'fieldsOfStudy',
              path: ['Demographics', 'Education', 'Fields of study', 'Business management'],
              name: 'Business management',
              audienceSize: 2636510.5,
            },
          ],
          educationStatuses: [{ type: 'educationStatuses', providerId: '' }],
        },
      },
      {
        or: {
          interests: [
            {
              providerId: '6002964301046',
              type: 'interests',
              path: ['Interests', 'Additional interests', 'Business school'],
              name: 'Business school',
              audienceSize: 41083805,
            },
          ],
          employers: [
            {
              providerId: '116050848438791',
              type: 'employers',
              path: ['Demographics', 'Work', 'Employers', 'Currently looking for a job.'],
              name: 'Currently looking for a job.',
              audienceSize: 3339.5,
            },
          ],
        },
      },
    ],
  },
  exclude: {
    or: {
      behaviors: [
        {
          providerId: '6071631541183',
          type: 'behaviors',
          path: ['Behaviours', 'Purchase behaviour', 'Engaged shoppers'],
          name: 'Engaged shoppers',
          audienceSize: 539317289.5,
        },
        {
          providerId: '123456',
          type: 'behaviors',
          path: ['Behaviours'],
          name: 'test',
          audienceSize: 123456,
        },
      ],
      demographics: [
        {
          providerId: '654321',
          type: 'demographics',
          path: ['Demographics'],
          name: 'test 2',
          audienceSize: 654321,
        },
      ],
      interests: [
        {
          providerId: '123654',
          type: 'interests',
          path: ['Interests'],
          name: 'test 2',
          audienceSize: 321654,
        },
      ],
    },
  },
};
