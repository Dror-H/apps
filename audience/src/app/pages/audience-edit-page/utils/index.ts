export * from './audience-rules/audience-rules.utils';
export * from './helpers/helpers.utils';
export * from './facebook-export-url/facebook-export-url';
