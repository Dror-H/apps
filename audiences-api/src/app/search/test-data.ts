export const mockSearchResults = [
  {
    id: '4870eab0-7ed2-56b1-9493-5b4684a7b03c',
    createdByContributors: false,
    spec: {
      id: '0478f7d5-55c7-4106-ba34-4f691236969c',
      name: '',
      targeting: {
        age_max: 65,
        age_min: 18,
        genders: [0],
        locales: [],
        exclusions: {},
        flexible_spec: [
          {
            work_positions: [
              {
                id: '105703889464434',
                name: 'Motivational speaker',
              },
              {
                id: '839841992748313',
                name: 'Personal Coach',
              },
            ],
          },
        ],
        geo_locations: {
          countries: ['US'],
          location_types: ['home', 'recent'],
        },
        excluded_geo_locations: {},
        targeting_optimization: 'none',
        excluded_custom_audiences: [],
      },
    },
    segments: [],
    userTags: [],
    topics: [],
  },
  {
    id: '478605de-ba43-5d8c-9c52-b1043e4cc0b7',
    createdByContributors: false,
    spec: {
      id: '23848984859650054',
      name: 'Interest Jewellry - UK - 18+ – Copy',
      campaign: {
        id: '23848984859400054',
        name: 'SO - Conversions - Bracelet - CBO - UK Only',
      },
      targeting: {
        age_max: 65,
        age_min: 35,
        genders: [2],
        exclusions: {
          interests: [
            {
              id: '6003109292833',
              name: 'Drop shipping',
            },
            {
              id: '6019154143076',
              name: 'AliExpress',
            },
          ],
        },
        flexible_spec: [
          {
            interests: [
              {
                id: '6003266225248',
                name: 'Jewellery',
              },
            ],
          },
        ],
        geo_locations: {
          countries: ['GB'],
          location_types: ['home', 'recent'],
        },
        brand_safety_content_filter_levels: ['FACEBOOK_STANDARD', 'AN_STANDARD'],
      },
    },
    segments: ['SO - Conversions - Bracelet - CBO - UK Only', 'Interest Jewellry - UK - 18+ – Copy', 'Jewellery'],
    userTags: [],
    topics: ['uk', 'jewellry', 'jewellery', 'interest', 'copy'],
  },
];

export const mockEmptyFlexSpecAndExclusions = {
  id: '478605de-ba43-5d8c-9c52-b1043e4cc0b7',
  createdByContributors: false,
  spec: {
    id: '23848984859650054',
    name: 'Interest Jewellry - UK - 18+ – Copy',
    campaign: {
      id: '23848984859400054',
      name: 'SO - Conversions - Bracelet - CBO - UK Only',
    },
    targeting: {
      age_max: 65,
      age_min: 35,
      genders: [2],
      exclusions: {
        interests: [],
        work_employers: [],
      },
      flexible_spec: [
        {
          interests: [],
          work_positions: [],
        },
      ],
      geo_locations: {
        countries: ['GB'],
        location_types: ['home', 'recent'],
      },
      brand_safety_content_filter_levels: ['FACEBOOK_STANDARD', 'AN_STANDARD'],
    },
  },
  segments: ['SO - Conversions - Bracelet - CBO - UK Only', 'Interest Jewellry - UK - 18+ – Copy', 'Jewellery'],
  userTags: [],
  topics: ['uk', 'jewellry', 'jewellery', 'interest', 'copy'],
};
