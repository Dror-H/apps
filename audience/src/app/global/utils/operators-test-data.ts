import { NotificationType, UINotification } from '@instigo-app/data-transfer-object';
import { ErrorType } from '@audience-app/global/utils/operators';

export interface AudiOperatorsTestCase<T, U = T> {
  case: T;
  expect: U | T;
}

function getBlankNotification(): UINotification {
  return {
    type: NotificationType.ERROR,
    content: undefined,
    title: undefined,
    options: undefined,
  };
}

export const notificationOptionsCases: AudiOperatorsTestCase<UINotification>[] = [
  {
    case: new UINotification({ type: NotificationType.ERROR, content: 'test' }),
    expect: new UINotification({ ...getBlankNotification(), content: 'test' }),
  },
  {
    case: new UINotification({
      type: NotificationType.ERROR,
      ...{ content: 'test', title: 'test title', options: {} },
    }),
    expect: new UINotification({
      ...getBlankNotification(),
      ...{ content: 'test', title: 'test title', options: {} },
    }),
  },
];
export const errorTypesTests: AudiOperatorsTestCase<ErrorType, UINotification>[] = [
  {
    case: { message: 'test message' },
    expect: new UINotification({ ...getBlankNotification(), content: 'test message' }),
  },
  {
    case: { error: { message: 'test error message' } },
    expect: new UINotification({ ...getBlankNotification(), content: 'test error message' }),
  },
  {
    case: { error: { description: 'test error description' } },
    expect: new UINotification({ ...getBlankNotification(), content: 'test error description' }),
  },
  {
    case: { error: { message: 'test error message', title: 'test error title' } },
    expect: new UINotification({
      ...getBlankNotification(),
      content: 'test error message',
      title: 'test error title',
    }),
  },
];
