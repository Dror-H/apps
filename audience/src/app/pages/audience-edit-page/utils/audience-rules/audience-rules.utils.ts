import { initFacebookTargetingRules } from '@audience-app/pages/audience-edit-page/mocks';
import { TargetingAndDto, TargetingDto } from '@instigo-app/data-transfer-object';
import { cloneDeep } from 'lodash';

export function createRulesFromTargeting(targeting?: TargetingDto): TargetingDto {
  if (!targeting) {
    return null;
  }

  const includeRules = getRulesWithDetailedTargeting(targeting);
  const targetingCopy = cloneDeep(targeting);
  targetingCopy.include.and = includeRules;

  return targetingCopy;
}

export function getRulesWithDetailedTargeting(targeting: TargetingDto): TargetingAndDto[] {
  const ands = targeting.include.and;
  const ageRange = ands[0].or;
  const locales = ands[1].or;
  const locationData = ands[2].or; //* can be non-existent

  const rules = initFacebookTargetingRules;
  if (locationData.locationTypes) {
    rules[0].or = locationData;
  }
  rules[1].or = ageRange;
  rules[3].or = locales;

  const detailedTargeting = getSelectedAudienceDetailedTargeting(targeting, !!locationData.locationTypes);
  return [...rules, ...detailedTargeting];
}

export function getSelectedAudienceDetailedTargeting(
  targeting: TargetingDto,
  hasLocationData: boolean,
): TargetingAndDto[] {
  return targeting.include.and.slice(hasLocationData ? 3 : 2);
}
