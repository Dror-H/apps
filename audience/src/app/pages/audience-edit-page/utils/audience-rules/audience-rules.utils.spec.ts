import * as funcs from './audience-rules.utils';
import { mockRegularTargeting, mockRegularTargetingWithLocation } from './test-data';

// todo - fix skipped tests

describe('audience rules utils', () => {
  describe('createRulesFromTargeting', () => {
    it('should return null if input is falsy', () => {
      expect(funcs.createRulesFromTargeting(null)).toBeNull();
    });

    it.skip('should call getRulesWithDetailedTargeting', () => {
      const spy = jest.spyOn(funcs, 'getRulesWithDetailedTargeting');
      funcs.createRulesFromTargeting(mockRegularTargeting);
      expect(spy).toBeCalledWith(mockRegularTargeting);
    });

    it('should create rules from targeting without location data', () => {
      const result = funcs.createRulesFromTargeting(mockRegularTargeting);
      expect(result).toMatchSnapshot();
    });

    it('should create rules from targeting with location data', () => {
      const result = funcs.createRulesFromTargeting(mockRegularTargetingWithLocation);
      expect(result).toMatchSnapshot();
    });
  });
  describe('getRulesWithDetailedTargeting', () => {
    it('should get rules with detailed targeting', () => {
      const result = funcs.getRulesWithDetailedTargeting(mockRegularTargeting);
      expect(result).toMatchSnapshot();
    });

    it.skip('should call getSelectedAudienceDetailedTargeting with input', () => {
      const spy = jest.spyOn(funcs, 'getSelectedAudienceDetailedTargeting');
      funcs.getRulesWithDetailedTargeting(mockRegularTargeting);
      expect(spy).toBeCalledWith(mockRegularTargeting);
    });
  });
  describe('getSelectedAudienceDetailedTargeting', () => {
    it('should get selected audience with detailed targeting', () => {
      const result = funcs.getRulesWithDetailedTargeting(mockRegularTargeting);
      expect(result).toMatchSnapshot();
    });
  });
});
