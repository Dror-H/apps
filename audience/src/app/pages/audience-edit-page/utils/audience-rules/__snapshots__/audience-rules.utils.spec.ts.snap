// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`audience rules utils createRulesFromTargeting should create rules from targeting with location data 1`] = `
Object {
  "exclude": Object {
    "or": Object {
      "behaviors": Array [
        Object {
          "audienceSize": 539317289.5,
          "name": "Engaged shoppers",
          "path": Array [
            "Behaviours",
            "Purchase behaviour",
            "Engaged shoppers",
          ],
          "providerId": "6071631541183",
          "type": "behaviors",
        },
        Object {
          "audienceSize": 123456,
          "name": "test",
          "path": Array [
            "Behaviours",
          ],
          "providerId": "123456",
          "type": "behaviors",
        },
      ],
      "demographics": Array [
        Object {
          "audienceSize": 654321,
          "name": "test 2",
          "path": Array [
            "Demographics",
          ],
          "providerId": "654321",
          "type": "demographics",
        },
      ],
      "interests": Array [
        Object {
          "audienceSize": 321654,
          "name": "test 2",
          "path": Array [
            "Interests",
          ],
          "providerId": "123654",
          "type": "interests",
        },
      ],
    },
  },
  "include": Object {
    "and": Array [
      Object {
        "or": Object {
          "countries": Array [],
          "locationTypes": Array [],
        },
      },
      Object {
        "or": Object {
          "ageRange": Array [
            Object {
              "providerId": "(25,50)",
              "type": "ageRange",
            },
          ],
        },
      },
      Object {
        "or": Object {
          "genders": Array [
            Object {
              "providerId": "0",
            },
          ],
        },
      },
      Object {
        "or": Object {
          "locales": Array [
            Object {
              "providerId": Array [],
            },
          ],
        },
      },
      Object {
        "or": Object {
          "targetOptimization": Array [
            Object {
              "providerId": false,
            },
          ],
        },
      },
      Object {
        "or": Object {
          "customAudiences": Array [
            Object {},
          ],
        },
      },
      Object {
        "or": Object {
          "facebookConnections": Array [
            Object {
              "providerId": Array [],
            },
          ],
        },
      },
      Object {
        "or": Object {
          "behaviors": Array [
            Object {
              "audienceSize": 48970322,
              "name": "Small business owners",
              "path": Array [
                "Behaviours",
                "Digital activities",
                "Small business owners",
              ],
              "providerId": "6002714898572",
              "type": "behaviors",
            },
          ],
          "employers": Array [
            Object {
              "audienceSize": 116699,
              "name": "Business Owner",
              "path": Array [
                "Demographics",
                "Work",
                "Employers",
                "Business Owner",
              ],
              "providerId": "109288035760640",
              "type": "employers",
            },
          ],
          "interests": Array [
            Object {
              "audienceSize": 192686100.5,
              "name": "Small business",
              "path": Array [
                "Interests",
                "Business and industry",
                "Small business",
              ],
              "providerId": "6002884511422",
              "type": "interests",
            },
          ],
        },
      },
      Object {
        "or": Object {
          "educationStatuses": Array [
            Object {
              "providerId": "",
              "type": "educationStatuses",
            },
          ],
          "fieldsOfStudy": Array [
            Object {
              "audienceSize": 2636510.5,
              "name": "Business management",
              "path": Array [
                "Demographics",
                "Education",
                "Fields of study",
                "Business management",
              ],
              "providerId": "106335129403847",
              "type": "fieldsOfStudy",
            },
          ],
          "interests": Array [
            Object {
              "audienceSize": 41083805,
              "name": "Business school",
              "path": Array [
                "Interests",
                "Additional interests",
                "Business school",
              ],
              "providerId": "6002964301046",
              "type": "interests",
            },
          ],
        },
      },
      Object {
        "or": Object {
          "employers": Array [
            Object {
              "audienceSize": 3339.5,
              "name": "Currently looking for a job.",
              "path": Array [
                "Demographics",
                "Work",
                "Employers",
                "Currently looking for a job.",
              ],
              "providerId": "116050848438791",
              "type": "employers",
            },
          ],
          "interests": Array [
            Object {
              "audienceSize": 41083805,
              "name": "Business school",
              "path": Array [
                "Interests",
                "Additional interests",
                "Business school",
              ],
              "providerId": "6002964301046",
              "type": "interests",
            },
          ],
        },
      },
    ],
  },
  "provider": "facebook",
}
`;

exports[`audience rules utils createRulesFromTargeting should create rules from targeting without location data 1`] = `
Object {
  "exclude": Object {
    "or": Object {
      "behaviors": Array [
        Object {
          "audienceSize": 539317289.5,
          "name": "Engaged shoppers",
          "path": Array [
            "Behaviours",
            "Purchase behaviour",
            "Engaged shoppers",
          ],
          "providerId": "6071631541183",
          "type": "behaviors",
        },
        Object {
          "audienceSize": 123456,
          "name": "test",
          "path": Array [
            "Behaviours",
          ],
          "providerId": "123456",
          "type": "behaviors",
        },
      ],
      "demographics": Array [
        Object {
          "audienceSize": 654321,
          "name": "test 2",
          "path": Array [
            "Demographics",
          ],
          "providerId": "654321",
          "type": "demographics",
        },
      ],
      "interests": Array [
        Object {
          "audienceSize": 321654,
          "name": "test 2",
          "path": Array [
            "Interests",
          ],
          "providerId": "123654",
          "type": "interests",
        },
      ],
    },
  },
  "include": Object {
    "and": Array [
      Object {
        "or": Object {
          "locationTypes": Array [],
        },
      },
      Object {
        "or": Object {
          "ageRange": Array [
            Object {
              "providerId": "(25,50)",
              "type": "ageRange",
            },
          ],
        },
      },
      Object {
        "or": Object {
          "genders": Array [
            Object {
              "providerId": "0",
            },
          ],
        },
      },
      Object {
        "or": Object {
          "locales": Array [
            Object {
              "providerId": Array [],
            },
          ],
        },
      },
      Object {
        "or": Object {
          "targetOptimization": Array [
            Object {
              "providerId": false,
            },
          ],
        },
      },
      Object {
        "or": Object {
          "customAudiences": Array [
            Object {},
          ],
        },
      },
      Object {
        "or": Object {
          "facebookConnections": Array [
            Object {
              "providerId": Array [],
            },
          ],
        },
      },
      Object {
        "or": Object {
          "behaviors": Array [
            Object {
              "audienceSize": 48970322,
              "name": "Small business owners",
              "path": Array [
                "Behaviours",
                "Digital activities",
                "Small business owners",
              ],
              "providerId": "6002714898572",
              "type": "behaviors",
            },
          ],
          "employers": Array [
            Object {
              "audienceSize": 116699,
              "name": "Business Owner",
              "path": Array [
                "Demographics",
                "Work",
                "Employers",
                "Business Owner",
              ],
              "providerId": "109288035760640",
              "type": "employers",
            },
          ],
          "interests": Array [
            Object {
              "audienceSize": 192686100.5,
              "name": "Small business",
              "path": Array [
                "Interests",
                "Business and industry",
                "Small business",
              ],
              "providerId": "6002884511422",
              "type": "interests",
            },
          ],
        },
      },
      Object {
        "or": Object {
          "educationStatuses": Array [
            Object {
              "providerId": "",
              "type": "educationStatuses",
            },
          ],
          "fieldsOfStudy": Array [
            Object {
              "audienceSize": 2636510.5,
              "name": "Business management",
              "path": Array [
                "Demographics",
                "Education",
                "Fields of study",
                "Business management",
              ],
              "providerId": "106335129403847",
              "type": "fieldsOfStudy",
            },
          ],
          "interests": Array [
            Object {
              "audienceSize": 41083805,
              "name": "Business school",
              "path": Array [
                "Interests",
                "Additional interests",
                "Business school",
              ],
              "providerId": "6002964301046",
              "type": "interests",
            },
          ],
        },
      },
      Object {
        "or": Object {
          "employers": Array [
            Object {
              "audienceSize": 3339.5,
              "name": "Currently looking for a job.",
              "path": Array [
                "Demographics",
                "Work",
                "Employers",
                "Currently looking for a job.",
              ],
              "providerId": "116050848438791",
              "type": "employers",
            },
          ],
          "interests": Array [
            Object {
              "audienceSize": 41083805,
              "name": "Business school",
              "path": Array [
                "Interests",
                "Additional interests",
                "Business school",
              ],
              "providerId": "6002964301046",
              "type": "interests",
            },
          ],
        },
      },
    ],
  },
  "provider": "facebook",
}
`;

exports[`audience rules utils getRulesWithDetailedTargeting should get rules with detailed targeting 1`] = `
Array [
  Object {
    "or": Object {
      "countries": Array [],
      "locationTypes": Array [],
    },
  },
  Object {
    "or": Object {
      "ageRange": Array [
        Object {
          "providerId": "(25,50)",
          "type": "ageRange",
        },
      ],
    },
  },
  Object {
    "or": Object {
      "genders": Array [
        Object {
          "providerId": "0",
        },
      ],
    },
  },
  Object {
    "or": Object {
      "locales": Array [
        Object {
          "providerId": Array [],
        },
      ],
    },
  },
  Object {
    "or": Object {
      "targetOptimization": Array [
        Object {
          "providerId": false,
        },
      ],
    },
  },
  Object {
    "or": Object {
      "customAudiences": Array [
        Object {},
      ],
    },
  },
  Object {
    "or": Object {
      "facebookConnections": Array [
        Object {
          "providerId": Array [],
        },
      ],
    },
  },
  Object {
    "or": Object {
      "behaviors": Array [
        Object {
          "audienceSize": 48970322,
          "name": "Small business owners",
          "path": Array [
            "Behaviours",
            "Digital activities",
            "Small business owners",
          ],
          "providerId": "6002714898572",
          "type": "behaviors",
        },
      ],
      "employers": Array [
        Object {
          "audienceSize": 116699,
          "name": "Business Owner",
          "path": Array [
            "Demographics",
            "Work",
            "Employers",
            "Business Owner",
          ],
          "providerId": "109288035760640",
          "type": "employers",
        },
      ],
      "interests": Array [
        Object {
          "audienceSize": 192686100.5,
          "name": "Small business",
          "path": Array [
            "Interests",
            "Business and industry",
            "Small business",
          ],
          "providerId": "6002884511422",
          "type": "interests",
        },
      ],
    },
  },
  Object {
    "or": Object {
      "educationStatuses": Array [
        Object {
          "providerId": "",
          "type": "educationStatuses",
        },
      ],
      "fieldsOfStudy": Array [
        Object {
          "audienceSize": 2636510.5,
          "name": "Business management",
          "path": Array [
            "Demographics",
            "Education",
            "Fields of study",
            "Business management",
          ],
          "providerId": "106335129403847",
          "type": "fieldsOfStudy",
        },
      ],
      "interests": Array [
        Object {
          "audienceSize": 41083805,
          "name": "Business school",
          "path": Array [
            "Interests",
            "Additional interests",
            "Business school",
          ],
          "providerId": "6002964301046",
          "type": "interests",
        },
      ],
    },
  },
  Object {
    "or": Object {
      "employers": Array [
        Object {
          "audienceSize": 3339.5,
          "name": "Currently looking for a job.",
          "path": Array [
            "Demographics",
            "Work",
            "Employers",
            "Currently looking for a job.",
          ],
          "providerId": "116050848438791",
          "type": "employers",
        },
      ],
      "interests": Array [
        Object {
          "audienceSize": 41083805,
          "name": "Business school",
          "path": Array [
            "Interests",
            "Additional interests",
            "Business school",
          ],
          "providerId": "6002964301046",
          "type": "interests",
        },
      ],
    },
  },
]
`;

exports[`audience rules utils getSelectedAudienceDetailedTargeting should get selected audience with detailed targeting 1`] = `
Array [
  Object {
    "or": Object {
      "countries": Array [],
      "locationTypes": Array [],
    },
  },
  Object {
    "or": Object {
      "ageRange": Array [
        Object {
          "providerId": "(25,50)",
          "type": "ageRange",
        },
      ],
    },
  },
  Object {
    "or": Object {
      "genders": Array [
        Object {
          "providerId": "0",
        },
      ],
    },
  },
  Object {
    "or": Object {
      "locales": Array [
        Object {
          "providerId": Array [],
        },
      ],
    },
  },
  Object {
    "or": Object {
      "targetOptimization": Array [
        Object {
          "providerId": false,
        },
      ],
    },
  },
  Object {
    "or": Object {
      "customAudiences": Array [
        Object {},
      ],
    },
  },
  Object {
    "or": Object {
      "facebookConnections": Array [
        Object {
          "providerId": Array [],
        },
      ],
    },
  },
  Object {
    "or": Object {
      "behaviors": Array [
        Object {
          "audienceSize": 48970322,
          "name": "Small business owners",
          "path": Array [
            "Behaviours",
            "Digital activities",
            "Small business owners",
          ],
          "providerId": "6002714898572",
          "type": "behaviors",
        },
      ],
      "employers": Array [
        Object {
          "audienceSize": 116699,
          "name": "Business Owner",
          "path": Array [
            "Demographics",
            "Work",
            "Employers",
            "Business Owner",
          ],
          "providerId": "109288035760640",
          "type": "employers",
        },
      ],
      "interests": Array [
        Object {
          "audienceSize": 192686100.5,
          "name": "Small business",
          "path": Array [
            "Interests",
            "Business and industry",
            "Small business",
          ],
          "providerId": "6002884511422",
          "type": "interests",
        },
      ],
    },
  },
  Object {
    "or": Object {
      "educationStatuses": Array [
        Object {
          "providerId": "",
          "type": "educationStatuses",
        },
      ],
      "fieldsOfStudy": Array [
        Object {
          "audienceSize": 2636510.5,
          "name": "Business management",
          "path": Array [
            "Demographics",
            "Education",
            "Fields of study",
            "Business management",
          ],
          "providerId": "106335129403847",
          "type": "fieldsOfStudy",
        },
      ],
      "interests": Array [
        Object {
          "audienceSize": 41083805,
          "name": "Business school",
          "path": Array [
            "Interests",
            "Additional interests",
            "Business school",
          ],
          "providerId": "6002964301046",
          "type": "interests",
        },
      ],
    },
  },
  Object {
    "or": Object {
      "employers": Array [
        Object {
          "audienceSize": 3339.5,
          "name": "Currently looking for a job.",
          "path": Array [
            "Demographics",
            "Work",
            "Employers",
            "Currently looking for a job.",
          ],
          "providerId": "116050848438791",
          "type": "employers",
        },
      ],
      "interests": Array [
        Object {
          "audienceSize": 41083805,
          "name": "Business school",
          "path": Array [
            "Interests",
            "Additional interests",
            "Business school",
          ],
          "providerId": "6002964301046",
          "type": "interests",
        },
      ],
    },
  },
]
`;
